### PR TITLE
overlord: implement store switch remodeling 

### PR DESCRIPTION
**Pull request by pedronis**
This is done using the corresponding remodel context.

So it starts using the remodel contexts in Remodel and doSetModel.

### CODE REVIEW

**Comment by pedronis** [comment]
based on #6838

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -666,15 +666,15 @@ func installGeneric(st *state.State, name, channel, cohort string, revision snap
 		return nil, errors.New("cannot specify revision and cohort")
 	}
 
-	return InstallUnderDeviceContext(st, name, channel, cohort, revision, userID, flags, nil)
+	return InstallWithDeviceContext(st, name, channel, cohort, revision, userID, flags, nil)
 }
 
-// InstallUnderDeviceContext returns a set of tasks for installing snap.
-// It will query for the snap under the given deviceCtx.
+// InstallWithDeviceContext returns a set of tasks for installing snap.
```
Comment (chipaca) [inline overlord/snapstate/snapstate.go]: *a* snap

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1263,15 +1263,15 @@ type RevisionOptions struct {
...
+	return UpdateWithDeviceContext(st, name, opts, userID, flags, nil)
 }
 
-// UpdateUnderDeviceContext initiates a change updating a snap.
-// It will query for the snap under the given deviceCtx.
+// UpdateWithDeviceContext initiates a change updating a snap.
+// It will query for the snap with the given deviceCtx.
 // Note that the state must be locked by the caller.
 //
 // The returned TaskSet will contain a DownloadAndChecksDoneEdge.
```
Comment (chipaca) [inline overlord/snapstate/snapstate.go]: I would've said it contains a `download-and-checks-done`

File: overlord/devicestate/devicestate.go [hunk 3]
```diff
@@ -456,21 +456,20 @@ func Remodel(st *state.State, new *asserts.Model) (*state.Change, error) {
...
-	if current.Model() != new.Model() {
-		return nil, fmt.Errorf("cannot remodel to different models yet")
-	}
-	if current.Store() != new.Store() {
-		return nil, fmt.Errorf("cannot remodel to different stores yet")
+
+	remodelKind := ClassifyRemodel(current, new)
+
+	if remodelKind == ReregRemodel {
+		return nil, fmt.Errorf("cannot remodel to different brand/model yet")
```
Comment (mvo5) [inline overlord/devicestate/devicestate.go]: :heart:

File: overlord/devicestate/devicestate.go [hunk 4]
```diff
@@ -505,7 +527,9 @@ func Remodel(st *state.State, new *asserts.Model) (*state.Change, error) {
 		msg = fmt.Sprintf(i18n.G("Remodel device to %v/%v (%v)"), new.BrandID(), new.Model(), new.Revision())
 	}
 	chg := st.NewChange("remodel", msg)
-	chg.Set("new-model", string(asserts.Encode(new)))
+	if err := remodCtx.Init(chg); err != nil {
```
Comment (mvo5) [inline overlord/devicestate/devicestate.go]: Nice seeing this coming together.

File: overlord/devicestate/devicestate_test.go [hunk 5]
```diff
@@ -2771,6 +2793,95 @@ func (s *deviceMgrSuite) TestRemodelLessRequiredSnaps(c *C) {
 	c.Assert(tSetModel.Summary(), Equals, "Set new model assertion")
 }
 
+type freshSessionStore struct {
```
Comment (mvo5) [inline overlord/devicestate/devicestate_test.go]: Nice name

File: overlord/devicestate/devicestate_test.go [hunk 6]
```diff
@@ -2771,6 +2793,95 @@ func (s *deviceMgrSuite) TestRemodelLessRequiredSnaps(c *C) {
...
+		"revision":       "1",
+	})
+
+	freshStore := &freshSessionStore{}
+	testStore = freshStore
+
+	s.newFakeStore = func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		mod, err := devBE.Model()
+		c.Check(err, IsNil)
+		if err == nil {
```
Comment (mvo5) [inline overlord/devicestate/devicestate_test.go]: (nitpick) We check "c.Check(err, IsNil)" before the "if err == nil" - can we remove the `if` or do you prefer it because it gives cleaner output on test failures? I think I slightly prefer it it would go away because it increase the cognitive load slightly when reading this.

File: overlord/devicestate/devicestate_test.go [hunk 7]
```diff
@@ -2771,6 +2793,95 @@ func (s *deviceMgrSuite) TestRemodelLessRequiredSnaps(c *C) {
...
+	}
+
+	chg, err := devicestate.Remodel(s.state, new)
+	c.Assert(err, IsNil)
+	c.Assert(chg.Summary(), Equals, "Refresh model assertion from revision 0 to 1")
+
+	c.Check(freshStore.ensureDeviceSession, Equals, 1)
+
+	tl := chg.Tasks()
+	// 2 snaps,
```
Comment (mvo5) [inline overlord/devicestate/devicestate_test.go]: (nitpick) Maybe: `// 2 snaps * 3 tasks (from the mock install above) + one  "set-model" task at the end` ?

File: overlord/devicestate/handlers.go [hunk 8]
```diff
@@ -122,9 +112,7 @@ func (m *DeviceManager) doSetModel(t *state.Task, _ *tomb.Tomb) error {
 		//       bootable base snap.
 	}
 
-	// TODO: set device,model from the new model assertion
-	// return setDeviceFromModelAssertion(st, device, model)
-	return nil
+	return remodCtx.Finish()
```
Comment (mvo5) [inline overlord/devicestate/handlers.go]: Nice!

File: overlord/devicestate/handlers_test.go [hunk 9]
```diff
@@ -89,6 +91,7 @@ func (s *deviceMgrSuite) TestSetModelHandlerSameRevisionNoError(c *C) {
 
 	t := s.state.NewTask("set-model", "set-model test")
 	chg := s.state.NewChange("dummy", "...")
+	chg.Set("new-model", string(asserts.Encode(model)))
 	chg.AddTask(t)
 	chg.Set("new-model", string(asserts.Encode(model)))
```
Comment (mvo5) [inline overlord/devicestate/handlers_test.go]: AFAICT we have `chg.Set("new-model", string(asserts.Encode(model)))` twice now - can we remove this line?

File: overlord/managers_test.go [hunk 10]
```diff
@@ -130,71 +131,77 @@ func verifyLastTasksetIsRerefresh(c *C, tts []*state.TaskSet) {
...
 	os.Setenv("SNAPPY_SQUASHFS_UNPACK_FOR_TESTS", "1")
 
 	// create a fake systemd environment
 	os.MkdirAll(filepath.Join(dirs.SnapServicesDir, "multi-user.target.wants"), 0755)
 
-	ms.restoreSystemctl = systemd.MockSystemctl(func(cmd ...string) ([]byte, error) {
+	r = systemd.MockSystemctl(func(cmd ...string) ([]byte, error) {
 		return []byte("ActiveState=inactive\n"), nil
 	})
+	s.AddCleanup(r)
```
Comment (stolowski) [inline overlord/managers_test.go]: +1 for migrating to AddClenaup in this test, thank you.

File: overlord/devicestate/handlers_test.go [hunk 11]
```diff
@@ -101,3 +104,67 @@ func (s *deviceMgrSuite) TestSetModelHandlerSameRevisionNoError(c *C) {
...
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+		"store":        "switched-store",
+		"revision":     "2",
+	})
+
+	s.newFakeStore = func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		mod, err := devBE.Model()
+		c.Check(err, IsNil)
+		if err == nil {
```
Comment (mvo5) [inline overlord/devicestate/handlers_test.go]: Same question as above about c.Check(err, IsNil) and if err == nil

File: overlord/managers_test.go [hunk 12]
```diff
@@ -335,15 +338,15 @@ func makeTestSnap(c *C, snapYamlContent string) string {
 	return snaptest.MakeTestSnapWithFiles(c, snapYamlContent, files)
 }
 
-func (ms *mgrsSuite) TestHappyLocalInstall(c *C) {
+func (s *mgrsSuite) TestHappyLocalInstall(c *C) {
```
Comment (stolowski) [inline overlord/managers_test.go]: The ms -> s renaming could be a separate PR to reduce the diff a bit.

File: overlord/devicestate/remodel.go [hunk 13]
```diff
@@ -266,6 +272,10 @@ type newStoreRemodelContext struct {
 	deviceMgr *DeviceManager
 }
 
+func (rc *newStoreRemodelContext) Kind() RemodelKind {
```
Comment (mvo5) [inline overlord/devicestate/remodel.go]: Thanks for this!

File: overlord/managers_test.go [hunk 14]
```diff
@@ -130,71 +131,77 @@ func verifyLastTasksetIsRerefresh(c *C, tts []*state.TaskSet) {
...
 
-	restoreConnectRetryTimeout := ifacestate.MockConnectRetryTimeout(connectRetryTimeout)
-
-	ms.restore = func() {
-		snapstate.SetupRemoveHook = oldSetupRemoveHook
-		snapstate.SetupInstallHook = oldSetupInstallHook
-		restoreBackendSave()
-		restoreConnectRetryTimeout()
-	}
+	s.AddCleanup(ifacestate.MockConnectRetryTimeout(connectRetryTimeout))
```
Comment (mvo5) [inline overlord/managers_test.go]: Nice!

File: overlord/managers_test.go [hunk 15]
```diff
@@ -130,71 +131,77 @@ func verifyLastTasksetIsRerefresh(c *C, tts []*state.TaskSet) {
...
 }
 
-func (ms *mgrsSuite) SetUpTest(c *C) {
-	ms.tempdir = c.MkDir()
-	dirs.SetRootDir(ms.tempdir)
+func (s *mgrsSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
+	s.tempdir = c.MkDir()
+	dirs.SetRootDir(s.tempdir)
```
Comment (mvo5) [inline overlord/managers_test.go]: (nipick) we *could* add a 
```go
s.AddCleanup(func() {dirs.SetRootDir("/")})
```
here and avoid it in the teardown.

File: overlord/managers_test.go [hunk 16]
```diff
@@ -130,71 +131,77 @@ func verifyLastTasksetIsRerefresh(c *C, tts []*state.TaskSet) {
...
-
-	ms.restore = func() {
-		snapstate.SetupRemoveHook = oldSetupRemoveHook
-		snapstate.SetupInstallHook = oldSetupInstallHook
-		restoreBackendSave()
-		restoreConnectRetryTimeout()
-	}
+	s.AddCleanup(ifacestate.MockConnectRetryTimeout(connectRetryTimeout))
 
 	os.Setenv("SNAPPY_SQUASHFS_UNPACK_FOR_TESTS", "1")
```
Comment (mvo5) [inline overlord/managers_test.go]: (nitpick) we *could* add the unset as a AddCleanup here and then we don't need an explict TearDown for this suite anymore afaict.

File: overlord/devicestate/devicestate.go [hunk 17]
```diff
@@ -505,7 +527,9 @@ func Remodel(st *state.State, new *asserts.Model) (*state.Change, error) {
...
+		return nil, err
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/devicestate/devicestate.go]: good point, I think all the current Init can only panic, not return an error. I should probably just remove the error return type from Init, need to double check with the last case that is WIP (rereg).

File: overlord/devicestate/handlers_test.go (hunk 9 above)
Comment (pedronis) [reply overlord/devicestate/handlers_test.go]: yes, probably some merge artifact
